Arrow X Datasets Blog (#1283)
huggingface/blog: arrow-datasets.md

---
title: "Data Analytics with Apache Arrow and Hugging Face Datasets" 
thumbnail: /blog/assets/arrow-datasets/bla.png
authors:
- user: cakiki
- user: lhoestq
- user: mariosasko
---

# Data Analytics with Apache Arrow and Hugging Face Datasets

<!-- {blog_metadata} -->
<!-- {authors} -->

> Hugging Face Datasets is a library that was designed to complement machine learning training workflows. It enables users to share, download, process, and format datasets to get them ready for training with any of the major machine learning frameworks, across all data modalities: audio, video, text, and tabular. In the Hugging Face ecosystem, the 🤗 `datasets` library is typically the first entrypoint of a machine learning project: it is therefore especially crucial it be performant and efficient.

Reviewer: ```suggestion
Hugging Face Datasets is a library that was designed to power machine learning training workflows. It enables users to share, download, process, and format datasets to get them ready for training with any of the major machine learning frameworks, across all data modalities: audio, video, text, and tabular. In the Hugging Face ecosystem, the 🤗 `datasets` library is typically the first entrypoint of a machine learning project: it is therefore especially crucial it be performant and efficient.
```

maybe? not sure

## Apache Arrow

According to the [official website](https://arrow.apache.org/), Apache Arrow is: (1) a column-oriented **standardized memory format**, (2) a set of **libraries** that implement said format and various utilities around it, and (3) an **ecosystem** of projects using them. 🤗 `datasets` is one of many projects of this ecosystem and leverages the Arrow memory format through the use of the [PyArrow library ](https://arrow.apache.org/docs/python/index.html): the official Python API of Apache Arrow. The Arrow format is further described as a:

> language-independent columnar memory format for flat and hierarchical data, organized for efficient analytic operations on modern hardware like CPUs and GPUs. The Arrow memory format also supports zero-copy reads for lightning-fast data access without serialization overhead.

Let's unpack this:

> - **language indepedent**: the data representation in memory is the same, no matter the programming language or library.

Reviewer: ```suggestion
- **language independent**: the data representation in memory is the same, no matter the programming language or library.
```

<!-- | ![Serialization](https://arrow.apache.org/img/copy.png) | ![Standardization](https://arrow.apache.org/img/shared.png)
|:--:|:--:|
|<i>TODO</i>|<i>TODO</i>|
<div align="center"> Source: <a href="https://arrow.apache.org/overview/" rel="noopener" target="_blank" >Apache Arrow Overview</a></div> -->

### Hugging Face Datasets

A Hugging Face `datasets.Dataset` object is at its core a dataframe-like structure called an Arrow Table: a 2-dimensional data structure with a schema and named columns. When you load a dataset for the first time, it is serialized in batches into one ore more `.arrow` memory files in your cache directory, then finally mmapped from those files on disk. To peek behind the scenes of this process let's begin by downloading the dataset we'll be using, a subset of [the Stack](https://huggingface.co/datasets/bigcode/the-stack-dedup), a dataset that was used to train the [StarCoder](https://huggingface.co/bigcode/starcoder) family of code large language models.

> ```python
> from datasets import load_dataset
> dset = load_dataset("cakiki/stack-smol-xxl", split="train") # ~107GB of disk space required to run this
> print(type(dset))
> print(type(dset.data.table)) # dset.data.table is the underlying Arrow Table
> print([shard["filename"].split("/")[-1] for shard in dset.cache_files])
> print(len(dset.cache_files), "cache files")
> >>> <class 'datasets.arrow_dataset.Dataset'> 
> >>> <class 'pyarrow.lib.Table'>
> >>> ['parquet-train-00000-of-00144.arrow', 'parquet-train-00001-of-00144.arrow', ... ] # These are the serialized files on disk. (Output trimmed) 
> >>> 144 cache files
> ```

Reviewer 2: Maybe reformat like this ? This would look like an ipython shell this way
````suggestion
```python
>>> from datasets import load_dataset
>>> dset = load_dataset("cakiki/stack-smol-xxl", split="train") # ~107GB of disk space required to run this
>>> type(dset)
<class 'datasets.arrow_dataset.Dataset'> 
>>> type(dset.data.table) # dset.data.table is the underlying Arrow Table
<class 'pyarrow.lib.Table'>
>>> [shard["filename"].split("/")[-1] for shard in dset.cache_files]
['parquet-train-00000-of-00144.arrow', 'parquet-train-00001-of-00144.arrow', ... ] # These are the serialized files on disk. (Output trimmed) 
>>> print(len(dset.cache_files), "cache files")
144 cache files
```
````

Caching onto disk is not limited to loading data, but also transforming it. Indeed, as we'll see in a second, when you operate on a dataset, be it through filtering, mapping, or structural transformations, these operations are also cached to disk and only need to run once. Before we move on to the next section, let's quickly inspect the code dataset we just loaded:

> ```python
> print(f'{dset.num_rows:,}', "rows")
> print(dset.num_columns, "columns")
> print(dset.dataset_size / 1024**3, " GB")
> print(dset[1200000])
> print(dset[:4])
> >>> 11,658,586 rows
> >>> 29 columns
> >>> 73.18686429131776 GB
> >>> {'hexsha': '7a4d22966630185a027c6ee50494472b3c1b3fa4', 'size': 10253, 'ext': 'jl', 'lang': 'Julia', 'max_stars_repo_path': 'src/cwv.jl' ...} # Indexing into a dataset returns the correspdonding row as a dictionary.
> >>> {'hexsha': ['0a84ade1a79baf9000df6f66818fc3a568c24173', '0a84b588e484c6829ab1a6706db302acd8514288', '0a89bca33bdfce1f07390d394915f18e0d901651', '0a96a3eeaa089f8582180f803252e3739ab2661d'], 'size': [10354, 999, 43147, 597], 'ext': ['css', 'css', 'css', 'css'], 'lang': ['CSS', 'CSS', 'CSS', 'CSS'], ...} # By default, slicing a dataset returns a dictionary of the columns as lists of regular python objects. Makes sense since arrow is a columnar format. (Output trimmed)
> ```

Reviewer 2: same here

````suggestion
```python
>>> print(f'{dset.num_rows:,}', "rows")
11,658,586 rows
>>> print(dset.num_columns, "columns")
29 columns
>>> print(f'{dset.dataset_size / 1024**3:.2f}', " GB")
73.18 GB
>>> dset[1200000]
'hexsha': '7a4d22966630185a027c6ee50494472b3c1b3fa4', 'size': 10253, 'ext': 'jl', 'lang': 'Julia', 'max_stars_repo_path': 'src/cwv.jl' ...} # Indexing into a dataset returns the correspdonding row as a dictionary.
>>> dset[:4]
{'hexsha': ['0a84ade1a79baf9000df6f66818fc3a568c24173', '0a84b588e484c6829ab1a6706db302acd8514288', '0a89bca33bdfce1f07390d394915f18e0d901651', '0a96a3eeaa089f8582180f803252e3739ab2661d'], 'size': [10354, 999, 43147, 597], 'ext': ['css', 'css', 'css', 'css'], 'lang': ['CSS', 'CSS', 'CSS', 'CSS'], ...} # By default, slicing a dataset returns a dictionary of the columns as lists of regular python objects. Makes sense since arrow is a columnar format. (Output trimmed)
```
````

### Hugging Face Dataset Formats

As you just saw, accessing data by slicing returns a dictionary where each column is a list of python objects, and accessing a single row returns a dictionary of the corresponding row. This dictionary format is the default way data is returned when accessing data  in the datasets library. This allows users to write pure Python `.map()` and `.filter()` operations that operate on python objects. 

## Mapping Compute Primitives

> ```python
> dset.set_format("arrow")
> print(dset[:4])
> >>> 
> pyarrow.Table
> hexsha: string
> size: int64
> ...
> ----
> hexsha: [["0a84ade1a79baf9000df6f66818fc3a568c24173","0a84b588e484c6829ab1a6706db302acd8514288","0a89bca33bdfce1f07390d394915f18e0d901651","0a96a3eeaa089f8582180f803252e3739ab2661d"]]
> size: [[10354,999,43147,597]]
> ext: [["css","css","css","css"]]
> lang: [["CSS","CSS","CSS","CSS"]]
> ...
> ```

Reviewer 2: ````suggestion
```python
>>> dset.set_format("arrow")
>>> print(dset[:4])
pyarrow.Table
hexsha: string
size: int64
...
----
hexsha: [["0a84ade1a79baf9000df6f66818fc3a568c24173","0a84b588e484c6829ab1a6706db302acd8514288","0a89bca33bdfce1f07390d394915f18e0d901651","0a96a3eeaa089f8582180f803252e3739ab2661d"]]
size: [[10354,999,43147,597]]
ext: [["css","css","css","css"]]
lang: [["CSS","CSS","CSS","CSS"]]
...
```
````

```python
print(pc.unique(dset['lang']).to_pylist())
>>> ['CSS', 'Prolog', 'C', 'FORTRAN', 'Solidity', 'Kotlin', 'Literate Agda', 'Julia', 'Java Server Pages', 'Isabelle', 'Idris', 'Lean', 'PowerShell', 'Go', 'Erlang', 'F#', 'Ada', 'Pascal', 'Perl', 'R', 'Protocol Buffer', 'CMake', 'SAS', 'Ruby', 'Rust', 'RMarkdown', 'C#', 'Smalltalk', 'Haskell', 'Maple', 'Mathematica', 'OCaml', 'Makefile', 'Lua', 'Literate CoffeeScript', 'Literate Haskell', 'reStructuredText', 'Racket', 'Standard ML', 'SystemVerilog', 'TeX', 'Awk', 'Assembly', 'Alloy', 'Agda', 'Emacs Lisp', 'Dart', 'Cuda', 'Bluespec', 'Augeas', 'Batchfile', 'Tcsh', 'Stan', 'Scala', 'Tcl', 'Stata', 'AppleScript', 'Shell', 'Clojure', 'Scheme', 'ANTLR', 'SPARQL', 'SQL', 'GLSL', 'Elm', 'Dockerfile', 'C++', 'CoffeeScript', 'Common Lisp', 'Elixir', 'Groovy', 'HTML', 'Java', 'JavaScript', 'Markdown', 'PHP', 'Python', 'TypeScript', 'Verilog', 'Visual Basic', 'VHDL', 'Thrift', 'Matlab', 'Yacc', 'Zig', 'XSLT', 'JSON', 'YAML']
CPU times: user 137 ms, sys: 1.65 ms, total: 138 ms
Wall time: 136 ms
```

```python

> mapped_result = dset.map(lambda table: table.group_by("lang").aggregate([("size", "mean"),("max_stars_count", "mean")]), batched=True, batch_size=100_000, num_proc=10)

Reviewer 2: would be cool to mention that converting to a pandas DataFrame to run this query would OOM, and only Dask would be able to run a similar query (not sure how fast it would be though ? we can leave this to readers)

Author: I agree it makes sense to mention both dask, and the new spark conversion in this context.

print(mapped_result)
>>> Dataset({
    features: ['size_mean', 'max_stars_count_mean', 'lang'],
    num_rows: 207
})
CPU times: user 765 ms, sys: 237 ms, total: 1 s

> Wall time: 1.35 s

Reviewer 2: This time is so impressive I think you can add a sentence to show what it represents in terms of GB/s

Reviewer 2: I guess you can include both if you want !

> ```python
> print(pc.unique(dset['lang']).to_pylist())
> >>> ['CSS', 'Prolog', 'C', 'FORTRAN', 'Solidity', 'Kotlin', 'Literate Agda', 'Julia', 'Java Server Pages', 'Isabelle', 'Idris', 'Lean', 'PowerShell', 'Go', 'Erlang', 'F#', 'Ada', 'Pascal', 'Perl', 'R', 'Protocol Buffer', 'CMake', 'SAS', 'Ruby', 'Rust', 'RMarkdown', 'C#', 'Smalltalk', 'Haskell', 'Maple', 'Mathematica', 'OCaml', 'Makefile', 'Lua', 'Literate CoffeeScript', 'Literate Haskell', 'reStructuredText', 'Racket', 'Standard ML', 'SystemVerilog', 'TeX', 'Awk', 'Assembly', 'Alloy', 'Agda', 'Emacs Lisp', 'Dart', 'Cuda', 'Bluespec', 'Augeas', 'Batchfile', 'Tcsh', 'Stan', 'Scala', 'Tcl', 'Stata', 'AppleScript', 'Shell', 'Clojure', 'Scheme', 'ANTLR', 'SPARQL', 'SQL', 'GLSL', 'Elm', 'Dockerfile', 'C++', 'CoffeeScript', 'Common Lisp', 'Elixir', 'Groovy', 'HTML', 'Java', 'JavaScript', 'Markdown', 'PHP', 'Python', 'TypeScript', 'Verilog', 'Visual Basic', 'VHDL', 'Thrift', 'Matlab', 'Yacc', 'Zig', 'XSLT', 'JSON', 'YAML']
> CPU times: user 137 ms, sys: 1.65 ms, total: 138 ms
> Wall time: 136 ms
> ```
> 
> ```python
> mapped_result = dset.map(lambda table: table.group_by("lang").aggregate([("size", "mean"),("max_stars_count", "mean")]), batched=True, batch_size=100_000, num_proc=10)
> print(mapped_result)
> >>> Dataset({
>     features: ['size_mean', 'max_stars_count_mean', 'lang'],
>     num_rows: 207
> })
> CPU times: user 765 ms, sys: 237 ms, total: 1 s
> Wall time: 1.35 s
> ```

Reviewer 2: ````suggestion
```python
>>> pc.unique(dset['lang']).to_pylist()
['CSS', 'Prolog', 'C', 'FORTRAN', 'Solidity', 'Kotlin', 'Literate Agda', 'Julia', 'Java Server Pages', 'Isabelle', 'Idris', 'Lean', 'PowerShell', 'Go', 'Erlang', 'F#', 'Ada', 'Pascal', 'Perl', 'R', 'Protocol Buffer', 'CMake', 'SAS', 'Ruby', 'Rust', 'RMarkdown', 'C#', 'Smalltalk', 'Haskell', 'Maple', 'Mathematica', 'OCaml', 'Makefile', 'Lua', 'Literate CoffeeScript', 'Literate Haskell', 'reStructuredText', 'Racket', 'Standard ML', 'SystemVerilog', 'TeX', 'Awk', 'Assembly', 'Alloy', 'Agda', 'Emacs Lisp', 'Dart', 'Cuda', 'Bluespec', 'Augeas', 'Batchfile', 'Tcsh', 'Stan', 'Scala', 'Tcl', 'Stata', 'AppleScript', 'Shell', 'Clojure', 'Scheme', 'ANTLR', 'SPARQL', 'SQL', 'GLSL', 'Elm', 'Dockerfile', 'C++', 'CoffeeScript', 'Common Lisp', 'Elixir', 'Groovy', 'HTML', 'Java', 'JavaScript', 'Markdown', 'PHP', 'Python', 'TypeScript', 'Verilog', 'Visual Basic', 'VHDL', 'Thrift', 'Matlab', 'Yacc', 'Zig', 'XSLT', 'JSON', 'YAML']
CPU times: user 137 ms, sys: 1.65 ms, total: 138 ms
Wall time: 136 ms
```

```python
>>> %time mapped_result = dset.map(lambda table: table.group_by("lang").aggregate([("size", "mean"),("max_stars_count", "mean")]), batched=True, batch_size=100_000, num_proc=10)
CPU times: user 765 ms, sys: 237 ms, total: 1 s
Wall time: 1.35 s
>>> mapped_result
Dataset({
    features: ['size_mean', 'max_stars_count_mean', 'lang'],
    num_rows: 207
})
```
````

> ```python
> reduced_result = mapped_result.map(lambda table: table.group_by("lang").aggregate([("size_mean", "mean"),("max_stars_count_mean", "mean")]), batched=True, batch_size=None)
> print(reduced_result.to_pandas().sort_values("size_mean_mean", ascending=False).head(15))
> >>>
> ```

Reviewer 2: ````suggestion
```python
>>> reduced_result = mapped_result.map(lambda table: table.group_by("lang").aggregate([("size_mean", "mean"),("max_stars_count_mean", "mean")]), batched=True, batch_size=None)
>>> reduced_result.to_pandas().sort_values("size_mean_mean", ascending=False).head(15)
```
````

|---:|-----------------:|----------------------------:|:------------|
| 30 |          63950.9 |                     62.9083 | Mathematica |
| 82 |          38485.2 |                   2032.26   | Matlab      |
| 86 |          19765.3 |                    173.915  | JSON        |
| 80 |          18590.4 |                     28.0008 | VHDL        |
|  9 |          17811.9 |                     33.3281 | Isabelle    |
| 68 |          16470.2 |                    118.346  | Common Lisp |
| 83 |          15776.3 |                     55.9424 | Yacc        |
| 71 |          15283.2 |                    294.813  | HTML        |
| 17 |          14399.4 |                     60.4394 | Pascal      |
| 22 |          13275   |                     22.6699 | SAS         |
| 85 |          13040.5 |                    170.301  | XSLT        |
| 55 |          13016.3 |                     34.9036 | Stata       |
| 62 |          12458.2 |                    339.297  | SQL         |
| 29 |          11677.5 |                    112.621  | Maple       |
| 84 |          11385.5 |                    349.611  | Zig         |

```python

> python = dset.map(lambda table: table.filter(pc.field("lang") == "Python"), batched=True, batch_size=500_000, num_proc=10)

Reviewer 3: It would be nice to compare this to the Python version (with lower `batch_size` to avoid OOM)

Author: Yes, good point! I'll add comparisons wherever it makes sense.

> ```python
> python = dset.map(lambda table: table.filter(pc.field("lang") == "Python"), batched=True, batch_size=500_000, num_proc=10)
> print(f'{python.num_rows:,}', "rows")
> >>> 250,000 rows
> CPU times: user 818 ms, sys: 255 ms, total: 1.07 s
> Wall time: 3.46 s
> ```

Reviewer 2: ````suggestion
```python
>>> %time python = dset.map(lambda table: table.filter(pc.field("lang") == "Python"), batched=True, batch_size=500_000, num_proc=10)
CPU times: user 818 ms, sys: 255 ms, total: 1.07 s
Wall time: 3.46 s
>>> print(f'{python.num_rows:,}', "rows")
250,000 rows
```
````

Reviewer 3: Maybe we can use the standard lib for measuring time instead of the magic commands so that this code works in the (standard) REPL

> ## Further Resources

Reviewer: Also mention Parquet as it's quite central to the Dataset Hub features? 

references: https://huggingface.co/blog/hub-duckdb
https://huggingface.co/docs/datasets-server/index

Reviewer: (and the link between Arrow and Parquet)

Author: Good point, I will add a snippet about parquet!

>>> print(len(dset.cache_files), "cache files")
144 cache files
```

> One might pause here and duly wonder whether this makes arrow a disk format. The data is after all sharded and saved on disk, and the answer is no. Seeing as the data is saved in the same format it exists in memory, the data columns aren't compressed which makes the format less suited for long term disk storage than its sister project: [**Apache Parquet**](https://parquet.apache.org/). Parquet is also a columnar data format, that focuses on efficient disk storage and retrieval, and it achieves this through efficiently compression each data column. Serializing from Parquet to Arrow and vice-versa is very easy, and the two projects are generally very closely related, so much so for example, that the C++ implementation of Parquet is now part of the Arrow project; but enough of this short interlude, and let's go back to our cached dataset.

Reviewer 2: Well it is a good disk format, it simply favours speed access with zero-copy than disk space ;)

```suggestion
The data is saved in the same format it exists in memory which allows fast access with zero-copy. However the data columns aren't compressed which makes the format less suited for long term disk storage than its sister project: [**Apache Parquet**](https://parquet.apache.org/). Parquet is also a columnar data format, that focuses on efficient disk storage and retrieval, and it achieves this through efficiently compression each data column. Serializing from Parquet to Arrow and vice-versa is very easy, and the two projects are generally very closely related, so much so for example, that the C++ implementation of Parquet is now part of the Arrow project; but enough of this short interlude, and let's go back to our cached dataset.
```
